### PR TITLE
Preview functionality for YAML config files

### PR DESCRIPTION
**Pull request by sihil**
This introduces the ability to preview deploys that use the new YAML configuration files.

**Note:** This does not provide any filtering yet.

Previews are still done in the same way from the deploy screen. They will now be sent to the new preview endpoint which will try to get the YAML config file. If it fails then it redirects the user to the old preview endpoint which will then get the JSON config.

Assuming it is a YAML config file then this is what it looks like:
screen shot 2016-10-21 at 18 25 21

Each dependency is an anchor link that jumps straight to the related tasks.

### CODE REVIEW

File: riff-raff/app/deployment/preview/Preview.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import com.gu.management.Loggable
+import magenta.`package`._
+import magenta.artifact.S3YamlArtifact
+import magenta.deployment_type.DeploymentType
+import magenta.graph.{DeploymentTasks, Graph}
+import magenta.input.resolver._
+import magenta.input.{ConfigErrors, Deployment}
+import magenta.{DeployParameters, DeploymentResources}
+
+object Preview extends Loggable {
```
Comment (philwills) [inline riff-raff/app/deployment/preview/Preview.scala]: Having different classes in the same project with the same name seems likely to cause confusion, even in different packages.

File: riff-raff/app/deployment/preview/Preview.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      previewGraph <- sequenceGraph {
+        flattenedGraph.map { deployment =>
+          TaskResolver.resolve(deployment, resources, parameters, deploymentTypes, artifact).map(deployment -> _)
+        }
+      }
+    } yield previewGraph
+    Preview(validatedGraph)
+  }
+
+  private[preview] def sequenceGraph[A, E](graph: Graph[Validated[E, A]])(implicit E: Semigroup[E]): Validated[E, Graph[A]] = {
```
Comment (philwills) [inline riff-raff/app/deployment/preview/Preview.scala]: Might it make sense to implement `Traverse` (http://typelevel.org/cats/typeclasses/traverse.html) for `Graph`?

File: riff-raff/app/deployment/preview/PreviewCoordinator.scala [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    val artifact = S3YamlArtifact.apply(parameters.build, conf.Configuration.artifact.aws.bucketName)
+    val maybeConfig = artifact.deployObject.fetchContentAsString()(resources.artifactClient)
+
+    maybeConfig match {
+      case Some(config) =>
+        logger.info(s"Got configuration for $previewId - resolving")
+        val previewFuture = Future(Preview(artifact, config, parameters, resources, deploymentTypes))
+        Await.ready(agent.alter{ _ + (previewId -> PreviewResult(previewFuture)) }, 30.second)
+        Right(previewId)
+      case None =>
```
Comment (philwills) [inline riff-raff/app/deployment/preview/PreviewCoordinator.scala]: There's a `toEither` in cats which I think saves a little noise for this kind of thing.

File: riff-raff/app/deployment/preview/PreviewCoordinator.scala [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import magenta.{DeployParameters, DeployReporter, DeploymentResources}
+import resources.PrismLookup
+
+import scala.concurrent.{Await, Future}
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+
+class PreviewCoordinator(prismLookup: PrismLookup, deploymentTypes: Seq[DeploymentType]) extends Loggable {
+  implicit lazy val system = ActorSystem("preview")
+  val agent = Agent[Map[UUID, PreviewResult]](Map.empty)
```
Comment (philwills) [inline riff-raff/app/deployment/preview/PreviewCoordinator.scala]: Might it be simpler to just use a `ConcurrentHashMap`?

File: riff-raff/app/deployment/preview/PreviewCoordinator.scala (hunk 3 above)
Comment (sihil) [inline riff-raff/app/deployment/preview/PreviewCoordinator.scala]: I can do a `toRight`? Or cats has a `fromOption` pimped on to the `Either` object? Do you mean one of those or something I've missed?

File: riff-raff/app/deployment/preview/PreviewCoordinator.scala (hunk 4 above)
Comment (sihil) [inline riff-raff/app/deployment/preview/PreviewCoordinator.scala]: Yes - that's panned out to be simpler.

File: riff-raff/app/deployment/preview/Preview.scala (hunk 1 above)
Comment (sihil) [inline riff-raff/app/deployment/preview/Preview.scala]: Yes - I have come to that conclusion whilst working on the filtering functionality! I've renamed the controller to `PreviewController` to make it obvious and have prefixed `Legacy` on the front of all the JSON preview classes.

File: riff-raff/app/deployment/preview/PreviewCoordinator.scala (hunk 3 above)
Comment (philwills) [inline riff-raff/app/deployment/preview/PreviewCoordinator.scala]: Ah, thought it was the former, but it's the latter:

https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/syntax/either.scala#L283

**Comment by adamnfish** [comment]
Is it missing a "sure, go ahead and deploy this" button?

**Pull request merged**